### ISSUE TITLE
Allow user to switch between languages & handle complexity of OptiKey language being different from OS language/keyboard layout

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#168]
*Story 1:* User needs to switch between languages from within OptiKey

**Chosen solution:** New LANGUAGES keyboard?

*Story 2:* When injecting letter/symbol/number keystrokes (using SendInput) we can either:

  1.Output UNICODE characters, which are always received as the output character, but are not detected by applications as physical keystrokes. This means shortcuts etc will not work.
  2.Output SCANCODES, which are very similar to sending VirtualKeyCodes, in that you need to know the physical layout of the keys + the current OS language/keyboard layout.
  3.Output VirtualKeyCodes, which are generally treated as "real" keyboard key presses, allowing text output and shortcuts etc.

**Chosen solution:**
I will attempt to calculate VK for each character using VkKeyScan and to the output add modifiers if required, or already down. If no VK returned fall back to outputting UNICODE. This allows users to combine modifiers with normal keys, but could result in extra keys which the user did not anticipate. Lots of logging required.
  +++ Log OS keyboard language/layout at startup and when it changes
  https://msdn.microsoft.com/en-us/library/windows/desktop/ms646329(v=vs.85).aspx
  https://msdn.microsoft.com/en-us/library/windows/desktop/ms646332(v=vs.85).aspx
```
        public static Keys ConvertCharToVirtualKey(char ch) {
            short vkey = VkKeyScan(ch);
            Keys retval = (Keys)(vkey & 0xff);
            int modifiers = vkey >> 8;
            if ((modifiers & 1) != 0) retval |= Keys.Shift;
            if ((modifiers & 2) != 0) retval |= Keys.Control;
            if ((modifiers & 4) != 0) retval |= Keys.Alt;
            return retval;
        }

        [System.Runtime.InteropServices.DllImport("user32.dll")]
        private static extern short VkKeyScan(char ch);
```
*N.B. It doesn't look like there is a solution which works in every scenario (Windows applications, DirectInput, DirectX, etc). There are other methods, such as keyboard driver wrappers (https://github.com/jasonpang/Interceptor), but these look like they only work in certain scenarios.*

*N.B.If OptiKey is failing for any particular user task then the user can still use the Windows OSK via mouse control.*

**Comment by Ar2rZ** [comment]
why don't leave as it is? as you said - there is no perfect solution for that any way.
1. add virtual key on and off function.
2. can be added additional chars to the foreign keyboards(to show english layout as it is on physical non-enlgish keyboards), and short keys will work perfectly, just will require to adjust virtual keys values  for each main language differently.
3. may be it will require to adjust input chars on scratchpad, to make consistent.
4. write into wiki that input language is depending on OS languages if virtual keys function is on.

so, may be i am thinking to simple ,there is something what i am missing?

**Comment by JuliusSweetland** [comment]
Hi Arthur. I might, but I'm experimenting at the moment. I think it's a little cleaner to ask Windows if there is a VirtualKeyCode combination that can achieve each character, and then output that combination. Also I'd like to detect the sent characters so I can update the scratchpad more correctly.

If it doesn't work well, or is too complicated then I'll go with your suggestion.

**Comment by JuliusSweetland** [comment]
New language keyboard and method of determining virtual key code released in v2.2.2
